File: back-end/src/application/services/publisher_service.rs
use crate::domain::entities::publisher::Publisher;
use crate::domain::irepositories::publisher_irepository::PublisherIRepository;
use async_trait::async_trait;

pub struct PublisherService<T: PublisherIRepository> {
    repository: T,
}

impl<T: PublisherIRepository> PublisherService<T> {
    pub fn new(repository: T) -> Self {
        Self { repository }
    }

    pub async fn register_publisher(&self, name: String, site: String, email: String, avatar: String) -> Result<(), String> {
        let publisher = Publisher::new(name, site, email, avatar);
        self.repository.create_publisher(&publisher).await
    }
}

File: front-end/src/aplication/services/PublisherService.ts
import { IPublisherRepository } from "../../domain/irepositories/IPublisherRepository";
import { Publisher } from "../../domain/entities/Publisher";


export class PublisherService implements IPublisherRepository {
    constructor(private PublisherRepository: IPublisherRepository) {}

    async getAllPublishers(): Promise<Publisher[]> {
        return await this.PublisherRepository.getAllPublishers();
    }

    async getPublisherById(id: number): Promise<Publisher> {
        return await this.PublisherRepository.getPublisherById(id);
    }

    async getPublisherByName(name: string): Promise<Publisher> {
        return await this.PublisherRepository.getPublisherByName(name);
    }

    async savePublisher(publisher: Publisher): Promise<void> {
        await this.PublisherRepository.savePublisher(publisher);
    }
}